docs: release notes for Charmcraft 4.0 (#2306)
canonical/charmcraft: docs/release-notes/charmcraft-4.0.rst

comes comes bundled with all its dependencies.

Non-snap installations of Charmcraft have the following dependencies:

- Python 3.10 or higher
- libgit2 1.7
- `skopeo`_
- `Spread`_

What's new
----------

Charmcraft 4.0 brings no new features, only removing deprecated features.

Feature removals
----------------

The following features are removed in Charmcraft 4.0. Anyone needing these features
can continue to use

Windows support
~~~~~~~~~~~~~~~

Charmcraft 4.0 is no longer tested at all on Windows. Windows users use
`Windows Subsystem for Linux <https://ubuntu.com/desktop/wsl>`_ to run Charmcraft.

CentOS 7 support
~~~~~~~~~~~~~~~~

CentOS 7 reached its `end of life
<https://www.redhat.com/en/topics/linux/centos-linux-eol>`_ on 30 June 2024, and
support has been removed from Charmcraft 4.0.

Bundle management commands
~~~~~~~~~~~~~~~~~~~~~~~~~~

Bundle management commands are being removed from Charmcraft 4.0. Charmcraft 3 will
continue to support bundles for those who still need them.

``bundle`` plugin
~~~~~~~~~~~~~~~~~

Charmcraft 4 does not pack bundles. Users who need to pack bundles can use Charmcraft 3
for this purpose.

``simple`` profile
~~~~~~~~~~~~~~~~~~

> The ``simple`` profile of ``charmcraft init`` has been removed from Charmcraft 4.
> The default profile is now ``kubernetes``, which is a minimal profile with scaffolding
> for a Kubernetes charm.
> Known issues

Reviewer: @lengau we just added the simple charm as an example in the Ops repo, in case anyone wants to continue using it for demos. What do you think about mentioning this in the release notes?

```suggestion
The ``simple`` profile of ``charmcraft init`` has been removed from Charmcraft 4.
The default profile is now ``kubernetes``, which is a minimal profile with scaffolding
for a Kubernetes charm. We have transferred the ``simple`` profile to an
`example charm <https://github.com/canonical/operator/tree/main/examples/httpbin-demo>`_
in the Ops repository.

Known issues
```

------------

The following issues were reported and are scheduled to be fixed in upcoming
patch releases.

See individual issue links for any mitigations.

- `#2078 <https://github.com/canonical/charmcraft/issues/2078>`_
  ``charmcraft clean`` does not clean all platforms for a charm.
- `#1990 <https://github.com/canonical/charmcraft/issues/1990>`_ Cannot stage
  packages with Charmcraft

Fixed bugs and issues
---------------------

> The following bugs have been resolved in Charmcraft 4.0.0:

Reviewer 2: I would say so. After having a look at our other projects, subsections seem to be the most up-to-date practice. 

[Charmcraft 3.4](https://canonical-charmcraft.readthedocs-hosted.com/3.5.2/release-notes/charmcraft-3.4/#fixed-bugs-and-issues)
[Snapcraft 8.10](https://documentation.ubuntu.com/snapcraft/stable/release-notes/snapcraft-8-10/#fixed-bugs-and-issues)

> The following bugs have been resolved in Charmcraft 4.0.0:

Reviewer 3: Yes please!

Charmcraft 4.0 release notes
============================

23 September 2025

Learn about the new features, changes, and fixes introduced in Charmcraft 4.0.

Requirements and compatibility
------------------------------

For development and testing, Charmcraft requires a host with a minimum of 4GB RAM
running a Linux distribution compatible with systemd.

All versions of Charmcraft require the following software:

- systemd
- `snapd`_
- Either `LXD`_ or `Multipass`_

We recommend you install the `Charmcraft snap <https://snapcraft.io/charmcraft>`_. It
comes comes bundled with all its dependencies.

Non-snap installations of Charmcraft have the following dependencies:

- Python 3.10 or higher
- libgit2 1.7
- `skopeo`_
- `Spread`_

What's new
----------

> Charmcraft 4.0 brings the following new features.

Author: Done!

OpenID Connect integration
~~~~~~~~~~~~~~~~~~~~~~~~~~

The App charm extensions now include OpenID Connect integration for app charms.
This applies automatically to any charm using the
:ref:`Django <django-framework-extension>`,
:ref:`Express <expressjs-framework-extension>`,
:ref:`FastAPI <fastapi-framework-extension>`,
:ref:`Flask <flask-framework-extension>`, and
:ref:`Go <go-framework-extension>` framework extensions when packing a charm with
Charmcraft 4.

``init`` profile changes
~~~~~~~~~~~~~~~~~~~~~~~~

The profiles for use with the :ref:`ref_commands_init` command have been modernised.

> The ``static`` and ``lint`` tox environments are no longer separate.

Author: Modified

Test profiles
^^^^^^^^^^^^^

Two new init profiles are also available for better integration with the
:ref:`ref_commands_test` command. These profiles are experimental, but are provided

> for early adopters who want to try the experimental test command.

Author: @medubelko @dwilding wdyt?

> for early adopters who want to try the experimental test command.

Reviewer: Looks good, thanks :slightly_smiling_face:

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Apps that use the Spring Boot framework can now use the
:ref:`spring-boot-framework-extension` for charming the app. A new
:ref:`tutorial <write-your-first-kubernetes-charm-for-a-spring-boot-app>` is also
included to walk a first-time charmer through charming their Spring Boot app.

Feature removals
----------------

The following features are removed in Charmcraft 4.0. Anyone needing these features
can continue to use

Windows support
~~~~~~~~~~~~~~~

Charmcraft 4.0 is no longer tested at all on Windows. Windows users use
`Windows Subsystem for Linux <https://ubuntu.com/desktop/wsl>`_ to run Charmcraft.

CentOS 7 support
~~~~~~~~~~~~~~~~

CentOS 7 reached its `end of life
<https://www.redhat.com/en/topics/linux/centos-linux-eol>`_ on 30 June 2024, and
support has been removed from Charmcraft 4.0.

Bundle management commands
~~~~~~~~~~~~~~~~~~~~~~~~~~

Bundle management commands are being removed from Charmcraft 4.0. Charmcraft 3 will
continue to support bundles for those who still need them.

``bundle`` plugin
~~~~~~~~~~~~~~~~~

Charmcraft 4 does not pack bundles. Users who need to pack bundles can use Charmcraft 3

> for this purpose.

Author: Done!

``simple`` profile
~~~~~~~~~~~~~~~~~~

The ``simple`` init profile is removed from Charmcraft 4.0.
The default profile is now ``kubernetes``, which is a minimal profile with scaffolding
for a Kubernetes charm. We have transferred the ``simple`` profile to an
`example charm <https://github.com/canonical/operator/tree/main/examples/httpbin-demo>`_
in the Ops repository.

Known issues
------------

The following issues were reported and are scheduled to be fixed in upcoming
patch releases.

See individual issue links for any mitigations.

- `#2078 <https://github.com/canonical/charmcraft/issues/2078>`_
  ``charmcraft clean`` does not clean all platforms for a charm.
- `#1990 <https://github.com/canonical/charmcraft/issues/1990>`_ Cannot stage
  packages with Charmcraft

Fixed bugs and issues
---------------------

Charmcraft 4.0.0
~~~~~~~~~~~~~~~~

- `#2431 <https://github.com/canonical/charmcraft/issues/2431>`_ — Charmcraft now
  errors if a relation name is invalid.

Contributors
------------

We would like to express a big thank you to all the people who contributed to
this release:

> :literalref:`@ade555<https://github.com/ade555>`,

Author: 🎉

- Python 3.10 or higher
- libgit2 1.7
- `skopeo`_
- `Spread`_

What's new
----------

Charmcraft 4.0 brings the following new features.

New Documentation Domain
~~~~~~~~~~~~~~~~~~~~~~~~

Charmcraft is an important member of the Canonical community and ecosystem, and all its
site names should demonstrate that. With this release, we gave the docs a new home at
the Ubuntu domain. You can now reach it at:

`documentation.ubuntu.com/charmcraft <https://documentation.ubuntu.com/charmcraft>`_

We put redirects in place to handle links to the old ReadTheDocs domain.

With this change, we also removed the language subdirectory (/en) in the URL, to
shave a few characters off all links.

OpenID Connect integration
~~~~~~~~~~~~~~~~~~~~~~~~~~

OpenID Connect is now available in all 12-factor app charm extensions.
It's available in any charm using the
:ref:`Django <django-framework-extension>`,
:ref:`Express <expressjs-framework-extension>`,
:ref:`FastAPI <fastapi-framework-extension>`,
:ref:`Flask <flask-framework-extension>`, and
:ref:`Go <go-framework-extension>` extensions and packed with Charmcraft 4.

``init`` profile changes
~~~~~~~~~~~~~~~~~~~~~~~~

> Charmcraft 4.0 has much more modern :ref:`ref_commands_init` ``machine`` and
> ``kubernetes`` profiles. Changes include:

Author: ```suggestion
Charmcraft 4.0 has much more modern ``machine`` and `kubernetes``
:ref:`ref_commands_init` profiles. Changes include:
```

?

> ``kubernetes`` profiles. Changes include:

Reviewer 3: @lengau's copy works for me.

- The ``static`` and ``lint`` tox environments are now united.
- The charm is packed with the :ref:`craft_parts_uv_plugin`.
- Dependencies are now managed with `uv <https://docs.astral.sh/uv>`_.

> 

Reviewer: Could we possibly add one more?

```suggestion
- Integration tests now use `Jubilant <https://documentation.ubuntu.com/jubilant/>`__

```

provide a framework for using the :ref:`charmcraft test <ref_commands_test>` command.
These profiles are experimental, so we can't guarantee their stability.

Spring Boot framework extension
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Apps that use the Spring Boot framework can now use the
:ref:`spring-boot-framework-extension` for charming the app. A new
:ref:`tutorial <write-your-first-kubernetes-charm-for-a-spring-boot-app>` is also
included to walk a first-time charmer through charming their Spring Boot app.

Feature removals
----------------

The following features are removed in Charmcraft 4.0. If you need these features, they
are still available in Charmcraft 3.

Windows support
~~~~~~~~~~~~~~~

Charmcraft 4.0 has ceased support for Windows. To continue on the platform, we
encourage Windows users to run Charmcraft in `Windows Subsystem for Linux
<https://ubuntu.com/desktop/wsl>`_.

CentOS 7 support
~~~~~~~~~~~~~~~~

CentOS 7 reached its `end of life
<https://www.redhat.com/en/topics/linux/centos-linux-eol>`_ on 30 June 2024, and
support for it has ended with Charmcraft 4.0.

Bundle removal
~~~~~~~~~~~~~~

> Aligning with the `discontinuation of bundle registrations
> <https://discourse.charmhub.io/t/15344>`_, Charmcraft 4.0 removes all bundle-related
> functionality. Charmcraft 3 will remain supported for packing bundles as well as the
> bundle commands (``register-bundle`` and ``promote-bundle``).

Reviewer 3: Sorry for the back and forth on this one, but I saw some ways we could be more attentive.

```suggestion
As scheduled, all charm bundle features are removed with Charmcraft 4.0. Charmcraft 3
will continue support for packing bundles and the ``register-bundle`` and
``promote-bundle`` commands. This follows the `discontinuation of new bundle
registrations <https://discourse.charmhub.io/t/15344>`_ in Nov 2024.
```

``simple`` profile
~~~~~~~~~~~~~~~~~~

The ``simple`` init profile is removed from Charmcraft 4.0.
The default profile is now ``kubernetes``, which is a minimal profile with scaffolding
for a Kubernetes charm. We have transferred the ``simple`` profile to an
`example charm <https://github.com/canonical/operator/tree/main/examples/httpbin-demo>`_
in the Ops repository.

Known issues
------------

The following issues were reported and are scheduled to be fixed in upcoming
patch releases.

See individual issue links for any mitigations.

- `#2078 <https://github.com/canonical/charmcraft/issues/2078>`_
  ``charmcraft clean`` does not clean all platforms for a charm.
- `#1990 <https://github.com/canonical/charmcraft/issues/1990>`_ Cannot stage
  packages with Charmcraft

Fixed bugs and issues
---------------------

Charmcraft 4.0.0
~~~~~~~~~~~~~~~~

> - `#2431 <https://github.com/canonical/charmcraft/issues/2431>`_ — Charmcraft now

Reviewer 3: Minor.

```suggestion
- `#2431 <https://github.com/canonical/charmcraft/issues/2431>`_ Charmcraft now
```

Charmcraft 4.0 release notes
============================

23 September 2025

Learn about the new features, changes, and fixes introduced in Charmcraft 4.0.

Requirements and compatibility
------------------------------

For development and testing, Charmcraft requires a host with a minimum of 4GB RAM
running a Linux distribution compatible with systemd.

All versions of Charmcraft require the following software:

- systemd
- `snapd`_
- Either `LXD`_ or `Multipass`_

We recommend you install the `Charmcraft snap <https://snapcraft.io/charmcraft>`_. It

> comes comes bundled with all its dependencies.

Reviewer 3: Mein Gott...

comes comes bundled with all its dependencies.

Non-snap installations of Charmcraft have the following dependencies:

- Python 3.10 or higher
- libgit2 1.7
- `skopeo`_
- `Spread`_

What's new
----------

Charmcraft 4.0 brings no new features, only removing deprecated features.

Feature removals
----------------

The following features are removed in Charmcraft 4.0. Anyone needing these features
can continue to use

Windows support
~~~~~~~~~~~~~~~

Charmcraft 4.0 is no longer tested at all on Windows. Windows users use
`Windows Subsystem for Linux <https://ubuntu.com/desktop/wsl>`_ to run Charmcraft.

CentOS 7 support
~~~~~~~~~~~~~~~~

CentOS 7 reached its `end of life
<https://www.redhat.com/en/topics/linux/centos-linux-eol>`_ on 30 June 2024, and
support has been removed from Charmcraft 4.0.

Bundle management commands
~~~~~~~~~~~~~~~~~~~~~~~~~~

Bundle management commands are being removed from Charmcraft 4.0. Charmcraft 3 will
continue to support bundles for those who still need them.

``bundle`` plugin
~~~~~~~~~~~~~~~~~

Charmcraft 4 does not pack bundles. Users who need to pack bundles can use Charmcraft 3
for this purpose.

> 

Reviewer: We'll need to mention the removal of `simple` (https://github.com/canonical/charmcraft/pull/2295):

```suggestion

``simple`` profile
~~~~~~~~~~~~~~~~~~

The ``simple`` profile of ``charmcraft ini`` has been removed from Charmcraft 4. The default profile is now ``kubernetes``, which is a minimal profile with scaffolding for a Kubernetes charm.
```

Author: Thanks David!
